src/models/clap.py

```diff
@@ -0,0 +1,38 @@
+import torch
+from .audio import BaseAudioEncoder
+from .text import BaseTextEncoder
+from typing import Union, List, Dict
+import numpy as np
+
+
+class BaseClap(torch.nn.Module):
+    def __init__(
+            self,
+            audio_encoder: BaseAudioEncoder,
+            text_encoder: BaseTextEncoder
+    ):
+        super().__init__()
+        self.audio_encoder = audio_encoder
+        self.text_encoder = text_encoder
+        self.loss_fn = torch.nn.CrossEntropyLoss()
+
+    def forward(self, texts: Union[List[str], str], audios: Union[List[np.array], np.array]) -> Dict[str, torch.Tensor]:
+        """
+        :param texts: texts for the input
+        :param audios: audios for the
+        :return: dictionary with cosine distance between embeddings
+            if len(texts) == len(audios) also returns the loss
+        """
+        num_texts = 1 if isinstance(texts, str) else len(texts)
+        num_audios = 1 if isinstance(audios, str) else len(audios)
+        texts_emb = self.text_encoder(texts)
+        audio_emb = self.audio_encoder(audios)
+        logits = texts_emb @ torch.transpose(audio_emb, 0, 1)
+        if num_texts == 1 or num_audios == 1:
+            logits = logits.squeeze()
+        ans = {"logits": logits}
+        if num_texts == num_audios and num_texts != 1:
```

Comment:
why we need num_texts == num_audios

Reply:
because otherwise you will have to have multiple audions generated from one theme, that creates some problems.
Also, I've checked, in the clip paper they trained with len(image_emb) == len(text_emb)